### PR TITLE
Stardew Valley: Generate proper filler for item links

### PR DESCRIPTION
**Pull request by agilbert1412**
## What is this fixing or adding?
It was brought to my attention that when playing with item links, Stardew Valley only generates Joja Cola. I fixed that to generate any resource pack. It can also provide traps if traps are enabled. Special conditional resource packs will only be provided if they should, based on settings

## How was this tested?
New unit tests for that feature, then running every test

### CODE REVIEW

**Comment by ProfBytes** [comment]
This should not generate traps.  If it does and replacement_link is on, it'll send traps to players with traps disabled.

**Comment by alwaysintreble** [comment]
> This should not generate traps. If it does and replacement_link is on, it'll send traps to players with traps disabled.

This is incorrect. The replacement items have to be the same for them to be linked.

File: worlds/stardew_valley/__init__.py [hunk 1]
```diff
@@ -267,7 +267,12 @@ def generate_basic(self):
         pass
 
     def get_filler_item_name(self) -> str:
-        return "Joja Cola"
+        include_traps = self.options[options.TrapItems] != options.TrapItems.option_no_traps
```
Comment (alwaysintreble) [inline worlds/stardew_valley/__init__.py:270]: This is an incredibly costly operation for how much this function can potentially get called. Should definitely cache the results. You inspired me to finally get around to doing this so example here: https://github.com/ArchipelagoMW/Archipelago/pull/2070

**Comment by ProfBytes** [comment]
> > This should not generate traps. If it does and replacement_link is on, it'll send traps to players with traps disabled.
> 
> This is incorrect. The replacement items have to be the same for them to be linked.

The documentation is wrong then.  The advanced yaml guide explicitly says that if link_replacement is true, they will be linked.  (I got the wrong order of the words in link replacement, so maybe that caused confusion)
This can be easily tested with a 2 game item link with one game having traps on and the other having them off.

File: worlds/stardew_valley/test/TestItemLink.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            if filler in filler_generated:
+                continue
+            filler_generated.append(filler)
+            self.assertNotIn(Group.MAXIMUM_ONE, item_table[filler].groups)
+            self.assertNotIn(Group.EXACTLY_TWO, item_table[filler].groups)
+            if Group.TRAP in item_table[filler].groups:
+                at_least_one_trap = True
+            if Group.GINGER_ISLAND in item_table[filler].groups:
+                at_least_one_island = True
+        print(len(filler_generated))
```
[the remark this answers is not in the document]
Comment (agilbert1412) [reply worlds/stardew_valley/test/TestItemLink.py]: Fixed

File: worlds/stardew_valley/__init__.py [hunk 3]
```diff
@@ -267,7 +268,30 @@ def generate_basic(self):
...
+        if self.player in self.multiworld.groups:
+            link_group: ItemLinkGroup = self.multiworld.groups[self.player]
+            include_traps = True
+            exclude_island = False
+            for player in link_group["players"]:
+                player_options = fetch_options(self.multiworld, player)
+                if player_options[options.TrapItems] == options.TrapItems.option_no_traps:
+                    include_traps = False
+                if player_options[options.ExcludeGingerIsland] == options.ExcludeGingerIsland.option_true:
+                    include_traps = exclude_island = True
```
[the remark this answers is not in the document]
Comment (agilbert1412) [reply worlds/stardew_valley/__init__.py]: Fixed

**Comment by ThePhar** [comment]
Has new merge conflicts now.
image

**Comment by agilbert1412** [comment]
> Has new merge conflicts now.

@ThePhar Fixed

File: worlds/stardew_valley/__init__.py [hunk 4]
```diff
@@ -270,7 +271,32 @@ def generate_basic(self):
         pass
 
     def get_filler_item_name(self) -> str:
-        return "Joja Cola"
+        if not self.filler_item_pool_names:
+            self.generate_filler_item_pool_names()
+        return self.multiworld.random.choice(self.filler_item_pool_names)
```
Comment (ThePhar) [inline worlds/stardew_valley/__init__.py:276]:
```suggestion
        return self.random.choice(self.filler_item_pool_names)
```

File: worlds/stardew_valley/__init__.py [hunk 5]
```diff
@@ -270,7 +271,32 @@ def generate_basic(self):
...
+        self.filler_item_pool_names = [item.name for item in available_filler]
+
+    def get_filler_item_rules(self):
+        if self.player in self.multiworld.groups:
+            link_group: ItemLinkGroup = self.multiworld.groups[self.player]
+            include_traps = True
+            exclude_island = False
+            for player in link_group["players"]:
+                player_options = self.multiworld.worlds[player].options
+                if not isinstance(player_options, StardewValleyOptions):
```
Comment (ThePhar) [inline worlds/stardew_valley/__init__.py:291]: Would it not be faster to just check the game of the player directly? Would be faster than a `isinstance` usage.

```suggestion
                if self.multiworld.game[player] != "Stardew Valley":

```

**Review by ThePhar — CHANGES_REQUESTED** [review]

File: worlds/stardew_valley/__init__.py (hunk 5 above)
Comment (agilbert1412) [reply worlds/stardew_valley/__init__.py]: Fair enough, I didn't think of it.

**Review by ThePhar — APPROVED** [review]

**Pull request merged**